File: backend/app/schemas/forecasts.py
"""Schemas for Traffic Forecasting API."""

from typing import List, Optional, Dict, Any
from pydantic import BaseModel
from backend.app.schemas.common import PaginationMeta, ProvenanceEnum


class ForecastRecord(BaseModel):
    """Multi-horizon forecast summary record."""
    timestamp: str
    segment_id: str
    current_speed_kmh: float
    pred_speed_15m: float
    pred_travel_time_15m_min: float
    pred_congestion_state_15m: str
    pred_speed_30m: float
    pred_speed_45m: float
    pred_speed_60m: float
    provenance: str = ProvenanceEnum.DERIVED.value


class SegmentForecastDetail(BaseModel):
    """Detailed forecast timeline with uncertainty bounds for a single segment."""
    timestamp: str
    segment_id: str
    current_speed_kmh: float
    horizons: Dict[str, Dict[str, Any]]
    provenance: str = ProvenanceEnum.DERIVED.value


class ForecastListResponse(BaseModel):
    """List of segment forecasts."""
    meta: PaginationMeta
    data: List[ForecastRecord]

File: frontend/types/forecasts.ts
/**
 * Forecast intelligence types matching FastAPI /api/forecasts response.
 */

import { PaginationMeta, ProvenanceType } from "./api";

export interface ForecastRecord {
  timestamp: string;
  segment_id: string;
  current_speed_kmh: number;
  pred_speed_15m: number;
  pred_travel_time_15m_min: number;
  pred_congestion_state_15m: string;
  pred_speed_30m: number;
  pred_speed_45m: number;
  pred_speed_60m: number;
  provenance: ProvenanceType | string;
}

export interface HorizonDetail {
  pred_speed_kmh: number;
  lower_bound_95_pct?: number;
  upper_bound_95_pct?: number;
  pred_travel_time_min?: number;
  pred_congestion_state: string;
}

export interface SegmentForecastDetail {
  timestamp: string;
  segment_id: string;
  current_speed_kmh: number;
  horizons: Record<string, HorizonDetail>;
  provenance: ProvenanceType | string;
}

export interface ForecastListResponse {
  meta: PaginationMeta;
  data: ForecastRecord[];
}

export interface ForecastQueryParams {
  limit?: number;
  offset?: number;
  segment_id?: string;
}
